src/vs/workbench/contrib/terminal/browser/terminalService.ts

```diff
@@ -836,17 +838,20 @@ export class TerminalService extends Disposable implements ITerminalService {
 	}
 
 	protected _initInstanceListeners(instance: ITerminalInstance): void {
-		const instanceDisposables: IDisposable[] = [
-			instance.onDimensionsChanged(() => {
-				this._onDidChangeInstanceDimensions.fire(instance);
-				if (this._terminalConfigurationService.config.enablePersistentSessions && this.isProcessSupportRegistered) {
-					this._saveState();
-				}
-			}),
-			instance.onDidFocus(this._onDidChangeActiveInstance.fire, this._onDidChangeActiveInstance),
-			instance.onRequestAddInstanceToGroup(async e => await this._addInstanceToGroup(instance, e))
-		];
-		this._register(instance.onDisposed(() => dispose(instanceDisposables)));
+		const instanceDisposables = new DisposableStore();
+		instanceDisposables.add(instance.onDimensionsChanged(() => {
+			this._onDidChangeInstanceDimensions.fire(instance);
+			if (this._terminalConfigurationService.config.enablePersistentSessions && this.isProcessSupportRegistered) {
+				this._saveState();
+			}
+		}));
+		instanceDisposables.add(instance.onDidFocus(this._onDidChangeActiveInstance.fire, this._onDidChangeActiveInstance));
+		instanceDisposables.add(instance.onRequestAddInstanceToGroup(async e => await this._addInstanceToGroup(instance, e)));
+		this._instanceDisposeListeners.set(instance, instanceDisposables);
+		this._register(instance.onDisposed(() => {
+			this._instanceDisposeListeners.delete(instance);
+			instanceDisposables.dispose();
+		}));
```

Comment:
There's still one more problem here where we need to get rid of `instance.onDisposed` when the instance goes down. `DisposableMap` was the suggestion for this, but looking at it again, I think we can do it simpler than that. We could remove `this._instanceDisposeListeners` and instead clean up the `instance.onDisposed` listener by doing something like this:

```ts
const disposeListener = this._register(instance.onDisposed(() => {
  instanceDisposables.dispose();
  this._store.delete(disposeListener);
});
```

That will delete the dispose listener for the no longer existing instance from the terminal service's `DisposableStore`.